projects/vdk-core/src/vdk/internal/builtin_plugins/run/data_job.py

```diff
@@ -253,6 +253,7 @@ def __init__(
             )
         self._name = data_job_directory.name if name is None else name
         self._data_job_directory = data_job_directory
+        core_context.state.set(ExecutionStateStoreKeys.JOB_NAME, self._name)
```

Comment:
This seems like a patch solution, and it may not address the root cause of the problem. 

If we do not address the root cause, it will surface in another way (with similar problem) . 

Let's think about what is the root cause? 

The context state is updated by [ExecutionTrackingPlugin](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/builtin_plugins/run/execution_tracking.py#L15) 
during initialize_job hook. 

While the [notifications](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/builtin_plugins/notification/notification.py#L115) are sent during vdk_exit (much later hook)

So why is not the state propagated properly? 

I think the issue might be in that , core_context.state is hierarchical. 

WHen a job is started a child context is created , and the job name is in the child context, then when the job finishes, teh child context is "poped" and the parent context knows nothing about the job state. 

I would change the [notification plugin](https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-core/src/vdk/internal/builtin_plugins/notification/notification.py#L115) to hook also into some hook during the [Job Run](https://github.com/vmware/versatile-data-kit/tree/main/projects/vdk-plugins#data-job-run-execution-cycle) and record as classic instance variable the needed information (job name, whatever else is needed) and use it during vdk_exit.

Reply:
I was hesitant to attach to yet another hook for this, as it seemed as too much of an overhead. But, yeah, I agree that in this case it is better to use a hook to avoid unexpected issues.